README.rst
Master Brick
============

This repository contains the firmware source code and the hardware design
files. The documentation generator configs can be found at
https://github.com/Tinkerforge/generators

Repository Content
------------------

software/:
 * examples/: Examples for all supported languages
 * build/: Compiled files
 * src/: Source code of firmware
 * Makefile: Makefile to build project

hardware/:
 * Contains KiCad project files and additionally schematics as PDF

datasheets/:
 * Contains datasheets for sensors and complex ICs that are used

Hardware
--------

The hardware is designed with the open source EDA Suite KiCad
(http://www.kicad.org). Before you are able to open the files,
you have to install the Tinkerforge kicad-libraries
(https://github.com/Tinkerforge/kicad-libraries). You can either clone
them directly in hardware/ or clone them in a separate folder and
symlink them into hardware/
(ln -s kicad_path/kicad-libraries project_path/hardware). After that you
can open the .pro file in hardware/ with KiCad and from there view and
modify the schematics and the PCB layout.

Software
--------

If you want to do your own Brick/Bricklet firmware development we highly
recommend that you use our build environment setup script and read the
tutorial (https://www.tinkerforge.com/en/doc/Tutorials/Tutorial_Build_Environment/Tutorial.html).

To compile the C code we recommend you to install the newest GNU Arm Embedded
Toolchain (https://launchpad.net/gcc-arm-embedded/+download).
You also need to install bricklib (https://github.com/Tinkerforge/bricklib).

You can either clone it directly in software/src/ or clone it in a
separate folder and symlink it into software/src/
(ln -s bricklib_path/bricklib project_path/software/src/). Finally make sure to
have CMake installed (http://www.cmake.org/cmake/resources/software.html).

After that you can build the firmware by invoking make in software/.
The firmware (.bin) can then be found in software/build/ and uploaded
with brickv (click button "Flashing" on start screen).


software/changelog
2012-02-15: 1.0.0 (e789418)
- Initial version

2012-02-21: 1.1.0 (a2de81e)
- Add support for tick types

2012-03-05: 1.1.1 (326879c)
- Fix bug with timeouts on Windows (through bricklib)

2012-03-11: 1.1.2 (4e32b76)
- Add USB request detection to allow Chibi slave stack to be powered over USB

2012-03-11: 1.1.3 (01d093b)
- Fix bug with ADC calibration readout (through bricklib)

2012-03-16: 1.1.4 (303a910)
- Use new heuristic for Chibi slave detection (master address == address)

2012-03-20: 1.1.5 (9852170)
- Disable IRQs in SPI master communication, otherwise Stepper Brick doesn't
  work over Chibi (through bricklib)

2012-03-22: 1.1.6 (c98f890)
- Fix several Chibi bugs:
  - Sanity check on Chibi participants discovery
  - If no ACK: slave recvs, master sends
  - No rx_aack_on state if Chibi recv buffer not empty

2012-03-29: 1.1.7 (957a138)
- Fix several Chibi bugs:
  - Check sequence number to eliminate duplicates
  - Different negotiation of send and recv states (now more messages from
    master to slave possible should be possible)

2012-06-27: 1.2.0 (d5a022d)
- Add RS485 support

2012-06-30: 1.2.1 (b936c3e)
- Add get_chip_temperature and reset function

2012-07-02: 1.2.2 (99509aa)
- Add USB hotplug support (through bricklib)

2012-07-24: 1.2.3 (72a6db9)
- Use dynamically calculated timeouts for RS485, improve Bricklet readout

2012-08-01: 1.2.4 (7dca31e)
- Send empty message if slave has timeout, to allow slave to process buffer

2012-09-18: 1.3.0 (a731acd)
- Add WIFI Extension support

2012-09-19: 1.3.1 (96f87c1)
- Remove unnecessary check (more bandwidth for WIFI Extension)

2012-09-20: 1.3.2 (35f8320)
- Add receive ringbuffer (prevent overflow due to callbacks)

2012-09-24: 1.3.3 (b0c36ea)
- Add support for ad-hoc and access point mode

2012-09-26: 1.3.4 (23e1a96)
- Add API for setting of WIFI regulatory domain

2012-10-04: 1.3.5 (bcd8582)
- Add GetUSBVoltage API
- Add autorestart after 3min without answer from GS1011
- add support for "," in WIFI passphrase

2012-10-04: 1.3.6 (e245cbc)
- Use WSEC=1 instead of auth mode = open for no encryption

2012-10-05: 1.4.0 (088c93d)
- Fix reset when multiple masters in a stack are used
- Add support for Master Brick 2.0

2012-10-07: 1.4.1 (8a1d82b)
- Fix bug in RS485 code (RS485 recv pin was set by WIFI code)

2012-10-11: 1.4.2 (578ea03)
- Add support for usage of two extensions at the same time

2012-10-16: 1.4.3 (2941a70)
- Fix mixup between ad-hoc DHCP and access point static IP

2012-10-16: 1.4.4 (2faed9e)
- Fix Chibi Extension initialization sequence

2012-10-18: 1.4.5 (f4f03f4)
- Fix Problem with RS485 + WIFI + Step Down PS combination

2012-10-26: 1.4.6 (cd9e0e5)
- Change Bricklet Context size from 250 to 256 (through bricklib)

2012-12-05: 2.0.0 (3710f26)
- Support for Protocol 2.0

2013-01-29: 2.0.1 (c96c28c)
- Fix buffer overflow in WIFI Extension code

2013-02-06: 2.0.2 (220a1c2)
- Fixes for a more stable WIFI Extension:
  - Remove command/data mode, use only one place for receiving/sending: allows
    for much better error handling, debugging and logging
  - Only save messages in brickd if return expected
  - Parse all command responses with human readable response
  - Parse <ESC>O, <ESC>F, <ESC>A options more thoroughly
  - Remove messages from brickd if socket disconnected
  - Make WIFI initialization asynchronous: allows to communicate with Master
    Brick while WIFI is connecting
  - Use 5 minute timeout before we assume a hang up (instead of 3)
  - Evaluate received data while sending command
  - Use different ATS configuration
  - Parse startup/reset message
  - Refactor data state machine

2013-02-07: 2.0.3 (03dedd5)
- Use smaller USB timeout on startup enumeration to allow faster enumeration
- Dont reread the config after the first WIFI association fails

2013-02-12: 2.0.4 (4b2c14a)
- Reset WIFI module after disassociate from access point
- Fix problem with changing of slave addresses of Chibi and RS485 Extension
- Externalize WIFI ringbuffer and feed bytes in state machine byte by byte
  (will allow the use of DMA controller later on)

2013-02-18: 2.0.5 (2728e6f)
- Reissue Low-Power-Mode after disassociate
- Add API for:
  - SetWifiHostname
  - GetWifiHostname
  - SetCurrentCallbackPeriod
  - GetCurrentCallbackPeriod
  - SetVoltageCallbackPeriod
  - GetVoltageCallbackPeriod
  - SetUSBVoltageCallbackPeriod
  - GetUSBVoltageCallbackPeriod
  - SetCurrentCallbackThreshold
  - GetCurrentCallbackThreshold
  - SetVoltageCallbackThreshold
  - GetVoltageCallbackThreshold
  - SetUSBVoltageCallbackThreshold
  - GetUSBVoltageCallbackThreshold
  - SetDebouncePeriod
  - GetDebouncePeriod
  - Current (Callback)
  - Voltage (Callback)
  - USBVoltage (Callback)
  - CurrentReached (Callback)
  - VoltageReached (Callback)
  - USBVoltageReached (Callback)

2013-04-04: 2.0.6 (44afc08)
- Add forced ACK after 10ms without answer
- Handle WIFI xon/xoff commands of GS1011MEES
- Add watchdog timer (restart after 16s of "deadlock")

2013-04-12: 2.1.0 (72d0099)
- Add support for Ethernet Extension

2013-06-06: 2.1.1 (cfac700)
- Increase Ethernet reset holding time from 2µs to 10µs to ensure proper bootup

2013-06-10: 2.1.2 (8699e40)
- Call Ethernet reset function 2 times to make sure it works always

2014-04-04: 2.2.0 (51d9e2a)
- Add support for Websockets for Ethernet Extension
- Add support for authentication of Ethernet/WIFI Extension
- Reset on USB resume to get a clean state
- Increase USB hotplug detect threshold
- SPI stack slave gives up immediately if message length too long

2014-04-09: 2.2.1 (679e62b)
- Remove "Suspend to Disk fix", seems to make problems on some windows versions
  (through bricklib)

2014-07-07: 2.2.2 (05aa82f)
- Remove USB hotplug support (through bricklib)

2014-11-25: 2.3.0 (60b3f17)
- Add support for new SPI protocol
- Add support for Master Brick 2.1

2015-02-20: 2.3.1 (baebb1a)
- Add support for Ethernet Extension 1.1
- Remove USB hotplug support for Master Brick 2.1 again

2015-05-20: 2.3.2 (b064927)

software/src/config.h
/* master-brick
 * Copyright (C) 2010-2015 Olaf Lüke <olaf@tinkerforge.com>
 *
 * config.h: Master-Brick specific configuration
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 */

#ifndef CONFIG_H
#define CONFIG_H

#include "bricklib/drivers/board/sam3s/SAM3S.h"
#include "master.h"
#include "communication.h"
#include "extensions/chibi/chibi.h"
#include "extensions/rs485/rs485.h"
#include "extensions/wifi/wifi.h"
#include "extensions/ethernet/ethernet.h"
#include "extensions/wifi2/wifi2.h"
#include "extensions/extension_config.h"

#define BRICK_FIRMWARE_VERSION_MAJOR 2
#define BRICK_FIRMWARE_VERSION_MINOR 5
#define BRICK_FIRMWARE_VERSION_REVISION 2

#define BRICK_HARDWARE_VERSION_MAJOR_32 3
#define BRICK_HARDWARE_VERSION_MINOR_32 2
#define BRICK_HARDWARE_VERSION_REVISION_32 0

#define BRICK_HARDWARE_VERSION_MAJOR_31 3
#define BRICK_HARDWARE_VERSION_MINOR_31 1
#define BRICK_HARDWARE_VERSION_REVISION_31 0

#define BRICK_HARDWARE_VERSION_MAJOR_21 2
#define BRICK_HARDWARE_VERSION_MINOR_21 1
#define BRICK_HARDWARE_VERSION_REVISION_21 0

#define BRICK_HARDWARE_VERSION_MAJOR_20 2
#define BRICK_HARDWARE_VERSION_MINOR_20 0
#define BRICK_HARDWARE_VERSION_REVISION_20 0

#define BRICK_HARDWARE_VERSION_MAJOR_10 1
#define BRICK_HARDWARE_VERSION_MINOR_10 0
#define BRICK_HARDWARE_VERSION_REVISION_10 0

#define BRICK_DEVICE_IDENTIFIER 13
#define BRICK_CAN_BE_MASTER
#define BRICK_ONLY_SUPPORTS_7P
#define BRICK_CAN_FLASH_BOOTLOADER

// ************** DEBUG SETTINGS **************
//#define DEBUG_SPI_STACK 1
//#define DEBUG_I2C_EEPROM 1
#define DEBUG_STARTUP 1
#define DEBUG_BRICKLET 1
#define DEBUG_CHIBI 1
#define DEBUG_MASTER 1
//#define PROFILING
//#define PROFILING_TIME 100 // After how many seconds profiling is printed

#define DISABLE_JTAG_ON_STARTUP
//#define LOGGING_SERIAL
//#define LOGGING_LEVEL LOGGING_DEBUG
#define LOGGING_LEVEL LOGGING_NONE

// ************** BRICK SETTINGS **************

// Frequencies
#define BOARD_MCK      64000000 // Frequency of brick
#define BOARD_MAINOSC  16000000 // Frequency of oscillator
#define BOARD_ADC_FREQ  6000000 // Frequency of ADC
#define BOARD_OSC_EXTERNAL      // Use external oscillator

#define BRICK_HAS_CO_MCU_SUPPORT

// UART for console output (printf)
//#define CONSOLE_USART_USE_UART1

#define PIN_CONSOLE_RXD  {1 << 21, PIOA, ID_PIOA, PIO_PERIPH_A, PIO_DEFAULT}
#define PIN_CONSOLE_TXD  {1 << 22, PIOA, ID_PIOA, PIO_PERIPH_A, PIO_DEFAULT}

#define CONSOLE_BAUDRATE 115200
#define CONSOLE_USART    USART1
#define CONSOLE_ID       ID_USART1
#define CONSOLE_PINS     {PIN_CONSOLE_RXD, PIN_CONSOLE_TXD}

#define PINS_UART        {PIN_CONSOLE_RXD, PIN_CONSOLE_TXD}

// TWI
// TWI version
#define TWI_V3XX

// TWI stack definitions (for reading of eeproms from Bricks in stack)
#define TWI_STACK           TWI1
#define PIN_TWI_TWD_STACK   {1 << 4, PIOB, ID_PIOB, PIO_PERIPH_A, PIO_DEFAULT}
#define PIN_TWI_TWCK_STACK  {1 << 5, PIOB, ID_PIOB, PIO_PERIPH_A, PIO_DEFAULT}
#define PINS_TWI_STACK      PIN_TWI_TWD_STACK, PIN_TWI_TWCK_STACK

#define PIN_SCL_PULLUP_MASTER   {1 << 0, PIOA, ID_PIOA, PIO_OUTPUT_1, PIO_DEFAULT}
#define PIN_SDA_PULLUP_MASTER   {1 << 17, PIOC, ID_PIOC, PIO_OUTPUT_1, PIO_DEFAULT}
#define PINS_TWI_PULLUP_MASTER  PIN_SCL_PULLUP_MASTER, PIN_SDA_PULLUP_MASTER

#define PIN_SCL_PULLUP_SLAVE    {1 << 0, PIOA, ID_PIOA, PIO_INPUT, PIO_DEFAULT}
#define PIN_SDA_PULLUP_SLAVE    {1 << 17, PIOC, ID_PIOC, PIO_INPUT, PIO_DEFAULT}
#define PINS_TWI_PULLUP_SLAVE   PIN_SCL_PULLUP_SLAVE, PIN_SDA_PULLUP_SLAVE

// TWI bricklet definitions (for bricklets, spi select and brick specific
//                           functions)
#define TWI_BRICKLET           TWI0
#define PIN_TWI_TWD_BRICKLET   {1 << 3, PIOA, ID_PIOA, PIO_PERIPH_A, PIO_DEFAULT}
#define PIN_TWI_TWCK_BRICKLET  {1 << 4, PIOA, ID_PIOA, PIO_PERIPH_A, PIO_DEFAULT}
#define PINS_TWI_BRICKLET      PIN_TWI_TWD_BRICKLET, PIN_TWI_TWCK_BRICKLET

// USB
// USB VBUS monitoring pin for USB plug and play
#define PIN_USB_DETECT      {1 << 17, PIOA, ID_PIOA, PIO_INPUT, PIO_DEFAULT}
#define PIN_MASTER20_DETECT {1 << 28, PIOC, ID_PIOC, PIO_INPUT, PIO_PULLUP}
#define PIN_MASTER21_DETECT {1 << 16, PIOC, ID_PIOC, PIO_INPUT, PIO_PULLUP}
#define PIN_MASTER31_DETECT {1 << 14, PIOC, ID_PIOC, PIO_INPUT, PIO_PULLUP}
#define PIN_MASTER32_DETECT {1 << 23, PIOA, ID_PIOA, PIO_INPUT, PIO_PULLUP}
// High if master in stack
#define PIN_3V3_ENABLE      {1 << 16, PIOA, ID_PIOA, PIO_OUTPUT_0, PIO_DEFAULT}


#define ADC_CAL_SUM 10

#define PIN_ADC_CAL_LOW     {1 << 3,  PIOB, ID_PIOB, PIO_INPUT, PIO_DEFAULT}
#define ADC_CAL_LOW_CHANNEL 7
#define ADC_CAL_LOW_MULTIPLIER 1
#define ADC_CAL_LOW_DIVISOR 11
#define ADC_CAL_LOW_REFERENCE 3300

#define PIN_ADC_CAL_HIGH    {1 << 15, PIOC, ID_PIOC, PIO_INPUT, PIO_DEFAULT}
#define ADC_CAL_HIGH_CHANNEL 11
#define ADC_CAL_HIGH_MULTIPLIER 10
#define ADC_CAL_HIGH_DIVISOR 11
#define ADC_CAL_HIGH_REFERENCE 3300


// USB product descriptor (name of brick)
#define PRODUCT_DESCRIPTOR { \
	USBStringDescriptor_LENGTH(12), \
    USBGenericDescriptor_STRING, \
    USBStringDescriptor_UNICODE('M'), \
    USBStringDescriptor_UNICODE('a'), \
    USBStringDescriptor_UNICODE('s'), \
    USBStringDescriptor_UNICODE('t'), \
    USBStringDescriptor_UNICODE('e'), \
    USBStringDescriptor_UNICODE('r'), \
    USBStringDescriptor_UNICODE(' '), \
    USBStringDescriptor_UNICODE('B'), \
    USBStringDescriptor_UNICODE('r'), \
    USBStringDescriptor_UNICODE('i'), \
    USBStringDescriptor_UNICODE('c'), \
    USBStringDescriptor_UNICODE('k') \
}


// SPI
#define PIN_SPI_MISO        {1 << 12, PIOA, ID_PIOA, PIO_PERIPH_A, PIO_DEFAULT}
#define PIN_SPI_MOSI        {1 << 13, PIOA, ID_PIOA, PIO_PERIPH_A, PIO_DEFAULT}
#define PIN_SPI_SPCK        {1 << 14, PIOA, ID_PIOA, PIO_PERIPH_A, PIO_DEFAULT}

#define PINS_SPI            PIN_SPI_MISO, PIN_SPI_MOSI, PIN_SPI_SPCK

#define PIN_SPI_SELECT_SLAVE  {1 << 11, PIOA, ID_PIOA, PIO_PERIPH_A, PIO_DEFAULT}

#define PIN_SPI_SELECT_MASTER_0 {1 << 20, PIOC, ID_PIOC, PIO_OUTPUT_1, PIO_DEFAULT}
#define PIN_SPI_SELECT_MASTER_1 {1 << 21, PIOC, ID_PIOC, PIO_OUTPUT_1, PIO_DEFAULT}
#define PIN_SPI_SELECT_MASTER_2 {1 << 22, PIOC, ID_PIOC, PIO_OUTPUT_1, PIO_DEFAULT}
#define PIN_SPI_SELECT_MASTER_3 {1 << 23, PIOC, ID_PIOC, PIO_OUTPUT_1, PIO_DEFAULT}
#define PIN_SPI_SELECT_MASTER_4 {1 << 0, PIOB, ID_PIOB, PIO_OUTPUT_1, PIO_DEFAULT}
#define PIN_SPI_SELECT_MASTER_5 {1 << 1, PIOB, ID_PIOB, PIO_OUTPUT_1, PIO_DEFAULT}
#define PIN_SPI_SELECT_MASTER_6 {1 << 2, PIOB, ID_PIOB, PIO_OUTPUT_1, PIO_DEFAULT}
#ifndef CONSOLE_USART_USE_UART1
#define PIN_SPI_SELECT_MASTER_7_20 {1 << 31, PIOC, ID_PIOC, PIO_OUTPUT_1, PIO_DEFAULT}
#define PIN_SPI_SELECT_MASTER_7_10 {1 << 3, PIOB, ID_PIOB, PIO_OUTPUT_1, PIO_DEFAULT}
#endif

#define PINS_SPI_SELECT_MASTER   PIN_SPI_SELECT_MASTER_0, \
                                 PIN_SPI_SELECT_MASTER_1, \
                                 PIN_SPI_SELECT_MASTER_2, \
                                 PIN_SPI_SELECT_MASTER_3, \
                                 PIN_SPI_SELECT_MASTER_4, \
                                 PIN_SPI_SELECT_MASTER_5, \
                                 PIN_SPI_SELECT_MASTER_6

// LED
#define PIN_LED_STD_BLUE    {1 << 18, PIOC, ID_PIOC, PIO_OUTPUT_1, PIO_DEFAULT}
#define PIN_LED_STD_RED     {1 << 19, PIOC, ID_PIOC, PIO_OUTPUT_1, PIO_DEFAULT}
#define PINS_STD_LED        PIN_LED_STD_BLUE, PIN_LED_STD_RED

#define PIN_LED_EXT_BLUE_0  {1 << 24, PIOC, ID_PIOC, PIO_OUTPUT_1, PIO_DEFAULT}
#define PIN_LED_EXT_BLUE_1  {1 << 25, PIOC, ID_PIOC, PIO_OUTPUT_1, PIO_DEFAULT}
#define PIN_LED_EXT_BLUE_2  {1 << 26, PIOC, ID_PIOC, PIO_OUTPUT_1, PIO_DEFAULT}
#define PIN_LED_EXT_BLUE_3  {1 << 27, PIOC, ID_PIOC, PIO_OUTPUT_1, PIO_DEFAULT}
#define PINS_EXT_LED        PIN_LED_EXT_BLUE_0, PIN_LED_EXT_BLUE_1, \
                            PIN_LED_EXT_BLUE_2, PIN_LED_EXT_BLUE_3

#define PINS_LED            PINS_STD_LED, PINS_EXT_LED

#define LED_STD_BLUE        0
#define LED_STD_RED         1
#define LED_EXT_BLUE_0      2
#define LED_EXT_BLUE_1      3
#define LED_EXT_BLUE_2      4
#define LED_EXT_BLUE_3      5

// Brick Detect
// Set low by master
// TODO: Change name
#define PIN_DETECT        {1 << 6, PIOB, ID_PIOB, PIO_OUTPUT_0, PIO_DEFAULT}

// If low, there is a master below (-> configure as stack participant)
#define PIN_MASTER_DETECT {1 << 18, PIOA, ID_PIOA, PIO_INPUT, PIO_PULLUP}


// ************** INTERRUPT PRIORITIES ***********
#define PRIORITY_EEPROM_MASTER_TWI0 0
#define PRIORITY_EEPROM_MASTER_TWI1 0
#define PRIORITY_EEPROM_SLAVE_TWI1 0
#define PRIORITY_STACK_SLAVE_SPI 0
#define PRIORITY_STACK_MASTER_SPI 0
#define PRIORITY_PROFILING_TC0 0


// ************** POWER MEASUREMENT **************
#define PIN_STACK_VOLTAGE {1 << 19, PIOA, ID_PIOA, PIO_INPUT, PIO_DEFAULT}
#define PIN_STACK_CURRENT {1 << 20, PIOA, ID_PIOA, PIO_INPUT, PIO_DEFAULT}

#define VOLTAGE_MAX_VALUE 4095

#define USB_VOLTAGE_CHANNEL 0
#define USB_VOLTAGE_MULTIPLIER (47+68)
#define USB_VOLTAGE_DIVISOR 68
#define USB_VOLTAGE_REFERENCE 3300

#define STACK_VOLTAGE_CHANNEL 2
#define STACK_VOLTAGE_MULTIPLIER 11
#define STACK_VOLTAGE_REFERENCE 3300

#define STACK_CURRENT_CHANNEL 3
#define STACK_CURRENT_MULTIPLIER 2
#define STACK_CURRENT_REFERENCE 3300

#define VOLTAGE_EPSILON 1000

// ************** BRICKLET SETTINGS **************

// Number of bricklet ports
#define BRICKLET_NUM 4

// BRICKLET A
#define BRICKLET_A_ADDRESS 84
#define BRICKLET_A_ADC_CHANNEL 12

#define BRICKLET_A_PIN_1_AD   {1 << 12, PIOC, ID_PIOC, PIO_INPUT, PIO_DEFAULT}
#define BRICKLET_A_PIN_2_DA   {1 << 25, PIOA, ID_PIOA, PIO_INPUT, PIO_DEFAULT}
#define BRICKLET_A_PIN_3_PWM  {1 << 0, PIOC, ID_PIOC, PIO_INPUT, PIO_DEFAULT}
#define BRICKLET_A_PIN_4_IO   {1 << 4, PIOC, ID_PIOC, PIO_INPUT, PIO_DEFAULT}
#define BRICKLET_A_PIN_SELECT {1 << 8, PIOC, ID_PIOC, PIO_OUTPUT_0, PIO_DEFAULT}


// BRICKLET B
#define BRICKLET_B_ADDRESS 84
#define BRICKLET_B_ADC_CHANNEL 10

#define BRICKLET_B_PIN_1_AD   {1 << 13, PIOC, ID_PIOC, PIO_INPUT, PIO_DEFAULT}
#define BRICKLET_B_PIN_2_DA   {1 << 26, PIOA, ID_PIOA, PIO_INPUT, PIO_DEFAULT}
#define BRICKLET_B_PIN_3_PWM  {1 << 1, PIOC, ID_PIOC, PIO_INPUT, PIO_DEFAULT}
#define BRICKLET_B_PIN_4_IO   {1 << 5, PIOC, ID_PIOC, PIO_INPUT, PIO_DEFAULT}
#define BRICKLET_B_PIN_SELECT {1 << 9, PIOC, ID_PIOC, PIO_OUTPUT_0, PIO_DEFAULT}


// BRICKLET C
#define BRICKLET_C_ADDRESS 84
#define BRICKLET_C_ADC_CHANNEL 13

#define BRICKLET_C_PIN_1_AD   {1 << 29, PIOC, ID_PIOC, PIO_INPUT, PIO_DEFAULT}
#define BRICKLET_C_PIN_2_DA   {1 << 13, PIOB, ID_PIOB, PIO_INPUT, PIO_DEFAULT}
#define BRICKLET_C_PIN_3_PWM  {1 << 2, PIOC, ID_PIOC, PIO_INPUT, PIO_DEFAULT}
#define BRICKLET_C_PIN_4_IO   {1 << 6, PIOC, ID_PIOC, PIO_INPUT, PIO_DEFAULT}
#define BRICKLET_C_PIN_SELECT {1 << 10, PIOC, ID_PIOC, PIO_OUTPUT_0, PIO_DEFAULT}


// BRICKLET D
#define BRICKLET_D_ADDRESS 84
#define BRICKLET_D_ADC_CHANNEL 14

#define BRICKLET_D_PIN_1_AD    {1 << 30, PIOC, ID_PIOC, PIO_INPUT, PIO_DEFAULT}
#define BRICKLET_D_PIN_1_AD_31 {1 << 7, PIOA, ID_PIOA, PIO_INPUT, PIO_DEFAULT}
#define BRICKLET_D_PIN_2_DA    {1 << 14, PIOB, ID_PIOB, PIO_INPUT, PIO_DEFAULT}
#define BRICKLET_D_PIN_2_DA_31 {1 << 4, PIOA, ID_PIOA, PIO_INPUT, PIO_DEFAULT}
#define BRICKLET_D_PIN_3_PWM   {1 << 3, PIOC, ID_PIOC, PIO_INPUT, PIO_DEFAULT}
#define BRICKLET_D_PIN_4_IO    {1 << 7, PIOC, ID_PIOC, PIO_INPUT, PIO_DEFAULT}
#define BRICKLET_D_PIN_SELECT  {1 << 11, PIOC, ID_PIOC, PIO_OUTPUT_0, PIO_DEFAULT}

// Available since HW version 3.0
#define PIN_BRICKLET_ENABLE {1 << 1, PIOA, ID_PIOA, PIO_OUTPUT_1, PIO_DEFAULT}

// ************** EXTENSION SETTINGS **************
#define PIN_EXT_0_SELECT    {1 << 28, PIOA, ID_PIOA, PIO_OUTPUT_0, PIO_DEFAULT}
#define PIN_EXT_0_SELECT_31 {1 << 29, PIOA, ID_PIOA, PIO_OUTPUT_0, PIO_DEFAULT}
#define PIN_EXT_0_GP_0_10   {1 << 9, PIOA, ID_PIOA, PIO_INPUT, PIO_DEFAULT}
#define PIN_EXT_0_GP_0_21   {1 << 8, PIOA, ID_PIOA, PIO_INPUT, PIO_DEFAULT}
#define PIN_EXT_0_GP_0_31   {1 << 10, PIOA, ID_PIOA, PIO_INPUT, PIO_DEFAULT}
#define PIN_EXT_0_GP_1      {1 << 10, PIOA, ID_PIOA, PIO_INPUT, PIO_DEFAULT}
#define PIN_EXT_0_GP_1_31   {1 << 8, PIOA, ID_PIOA, PIO_INPUT, PIO_DEFAULT}
#define PIN_EXT_0_GP_2      {1 << 27, PIOA, ID_PIOA, PIO_INPUT, PIO_DEFAULT}
#define PIN_EXT_0_GP_2_31   {1 << 8, PIOC, ID_PIOC, PIO_INPUT, PIO_DEFAULT}

#define PIN_EXT_1_SELECT    {1 << 29, PIOA, ID_PIOA, PIO_OUTPUT_0, PIO_DEFAULT}
#define PIN_EXT_1_SELECT_31 {1 << 28, PIOA, ID_PIOA, PIO_OUTPUT_0, PIO_DEFAULT}
#define PIN_EXT_1_GP_0      {1 << 1, PIOA, ID_PIOA, PIO_INPUT, PIO_DEFAULT}
#define PIN_EXT_1_GP_0_31   {1 << 30, PIOA, ID_PIOA, PIO_INPUT, PIO_DEFAULT}
#define PIN_EXT_1_GP_1      {1 << 30, PIOA, ID_PIOA, PIO_INPUT, PIO_DEFAULT}
#define PIN_EXT_1_GP_1_31   {1 << 3, PIOA, ID_PIOA, PIO_INPUT, PIO_DEFAULT}
#define PIN_EXT_1_GP_2      {1 << 31, PIOA, ID_PIOA, PIO_INPUT, PIO_DEFAULT}
#define PIN_EXT_1_GP_2_31   {1 << 27, PIOA, ID_PIOA, PIO_INPUT, PIO_DEFAULT}

#define PINS_EXT_10       PIN_EXT_0_GP_0_10, PIN_EXT_0_GP_1, PIN_EXT_0_GP_2, \
                          PIN_EXT_1_GP_0, PIN_EXT_1_GP_1, PIN_EXT_1_GP_2
#define PINS_EXT_21       PIN_EXT_0_GP_0_21, PIN_EXT_0_GP_1, PIN_EXT_0_GP_2, \
                          PIN_EXT_1_GP_0, PIN_EXT_1_GP_1, PIN_EXT_1_GP_2
#define PINS_EXT_31       PIN_EXT_0_GP_0_21, PIN_EXT_0_GP_1, PIN_EXT_0_GP_2_31, \
                          PIN_EXT_1_GP_0_31, PIN_EXT_1_GP_1_31, PIN_EXT_1_GP_2_31


#define PIN_EXT_RXD       {1 << 5, PIOA, ID_PIOA, PIO_INPUT, PIO_DEFAULT}
#define PIN_EXT_TXD       {1 << 6, PIOA, ID_PIOA, PIO_INPUT, PIO_DEFAULT}
#define PIN_EXT_SCK       {1 << 2, PIOA, ID_PIOA, PIO_INPUT, PIO_DEFAULT}

// ************** MASTER DEBUGGING **************
#define PIN_STACK_SYNC    {1 << 7, PIOB, ID_PIOB, PIO_OUTPUT_0, PIO_DEFAULT}

#if(DEBUG_MASTER)
#define logmasterd(str, ...) do{logd("master: " str, ##__VA_ARGS__);}while(0)
#define logmasteri(str, ...) do{logi("master: " str, ##__VA_ARGS__);}while(0)
#define logmasterw(str, ...) do{logw("master: " str, ##__VA_ARGS__);}while(0)
#define logmastere(str, ...) do{loge("master: " str, ##__VA_ARGS__);}while(0)
#define logmasterf(str, ...) do{logf("master: " str, ##__VA_ARGS__);}while(0)
#else
#define logmasterd(str, ...) {}
#define logmasteri(str, ...) {}
#define logmasterw(str, ...) {}
#define logmastere(str, ...) {}
#define logmasterf(str, ...) {}
#endif



#endif
